[Week5] Customer & Validation

sunggyupaik opened the pull request:
# 요구사항

## REST API
- [x] 1. 회원 생성하기 - POST /user
- [x] 2. 회원 수정하기 - POST /user/{id}
- [x] 3. 회원 삭제하기 - DELETE /user/{id}

## 테스트 검사
- [x] 1.  회원 생성하기 - POST /user
- [x] 2. 회원 수정하기 - PATCH /user/{id}
- [x] 3. 회원 삭제하기 - DELETE /user/{id}

## 유효성 검사
- [x] 1. 고양이 장난감가게 유효성 검사 추가하기
- [x] 2. 회원 관리하기

### 2021-02-22 ( 1일차 )
 - 강의보며 이해하기

### 2021-02-23 ( 2일차 )
 - 강의보며 이해하기
 - Product Validation 추가
 - 
### 2021-02-24 ( 3일차 )
 - `@NotNull`, `@NotBlank`
 - `@Valid` 처리
 - Product Validation 추가
 - `User`에 대한 테스트 추가

### 2021-02-25 ( 4일차 )
- `UserData` 추가 및 구현
 - `ErrorAdvice` 예외처리 변경하기
 - 테스트 `setUp` , `Builder` 추가
 - 예외상황에 대한 테스트 추가
 - JavaDoc 주석달기
 - `Dozrer` 공부, 사용
 
 
## 추가사항


## 개선사항

sunggyupaik committed:
Add test about valid, bad request

sunggyupaik committed:
Make POST /user controller test

sunggyupaik committed:
Make fast User class

sunggyupaik committed:
Add given, content at POST /users controllertest

sunggyupaik commented:
이번 한주간도 잘 부탁드립니다 ㅎㅎ

sunggyupaik commented:
오늘 좀 열심히 들여다보면서 궁금증이 몇가지 생겼습니다. 시간 상 더 깊이 찾아보지 못했는지 인사이트 주시면 감사하겠습니다!!

1. `Controller`에서 `@Valid`에 대한 예외 처리가 `MethodArgumentNotValidException.class`로 가능한 걸로 알고 있습니다. 그런데, 저는 이전에 우리가 NOT_FOUND 예외를 `ProductNotFoundException `으로 처리했던 것처럼 잘못된 요청에 대해서는`ProductBadReqeustException`을 통해서 처리하기를 원합니다. 제가 만든 예외로 커스터마이징해서 인식하게 하려면 어떻게 해야 할까요?

2. `Controller`에서 `@Valid`가 처리되는 것으로 보아, `Service`에서는 `@Valid`와 연관이 별로 없는 것으로 생각됩니다. 이게 맞다면, `ServiceTest`에서는 가령 `create`에서 이름이 없다던가, 메이커가 없다든가 등의 잘못된 요청에 대한 예외 작성은 어떤 식으로 해야할까요? `Service` 내에서는 예외를 잡는 부분이 없는 것 같아서 `given`, `verify`에 대한 작성을 못했습니다.

3-1.  길어지는 update에서의 `@Display`를 어떻게 작성해야 할지 고민입니다.
```
@DisplayName("만약 저징되어 있는 고양이 장난감의 아이디와 업데이트 될 이름, 메이커, 가격, 이미지가 주어진다면")
```
이것이 명확하게 동작을 드러낸다고 생각했는데 너무 긴 것이 아닌가 생각했습니다.
```
@DisplayName("만약 저장되어 있는 고양이 장난감의 아이디와 업데이트 될 객체가 주어진다면")
```

3-2. 업데이트 시, 매개변수는 `id`와 `Product`가 모두 필요한대, id와 상관없이 객체에서 `메이커가 없을 때` 를 검사한다면, id에 대한 부분은 생략해도 될까요?
```
@DisplayName("만약 저징되어 있는 고양이 장난감의 아이디와 메이커가 비어있는 객체가 주어진다면")
```

```
@DisplayName("만약 메이커가 비어있는 객체가 주어진다면")
```


4. `@NotNull` , `@NotBlank` 의 어노테이션에 대한 테스트는 어떻게 면밀히 검증할까요?
`@NotNull`의 경우 `null`은 불가, `""`와 `" "`는 허용
`@NotBlank`의 경우 `null`, `""`, `" "`모두 불가

`null`, `""`, `" "` 모두 유의미한 차이가 있을까요? 먼저, 입력값이 없으면 `생성` 버튼을 눌렀을 때, null이 넘어가나요, `""`이 넘어가나요? `@NotNull`과 `@NotBlank`라고 구분해서 사용한 것이 유의미한 차이가 있을거라고 생각해서 `@DisplayName`에서 `blank 일 때`, `null일 때` 이런식으로 나눴는데, 크게 상관없을까요?

5. `jacoco`에서 `ToString()`에 대한 부분이 계속 빨간색이 뜨는데, 여기서 원하는 처리는 어떤 것일까요?

johngrib commented:
### `@Valid`
https://github.com/edalorzo/fundamental-programming#spring-controller-barricade
이런 방법이 있습니다. 그런데 실제로 쓰는 건 거의 못 봤어요. 보통 `@Valid`의 처리 결과 던져진 예외는 `@NotBlank` 같은 애노테이션에 인자로 넘겨준 에러 메시지로 응답하도록 작업해 둡니다. 그래서 `@NotBlank(message = "공백이면 안됩니다.")` 보통 이렇게 하는 것으로 충분한 느낌입니다.

현재 작업하신 상태에서 제가 다음과 같이 작업해 봤습니다.

| `@NotBlank` | `ErrorAdvice` |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1855714/108852282-bf964d80-7628-11eb-9465-94ab23e22aa8.png) | ![image](https://user-images.githubusercontent.com/1855714/108852311-ca50e280-7628-11eb-971e-f36aef6c60f4.png) |

그리고 다음과 같이 `curl`을 사용해 HTTP 요청을 보내봤습니다.
```
curl -s -H 'Content-Type: application/json' \
  -D - -X POST \
  --data '{  "name": "123", "maker": "", "price": 1}' 'http://localhost:8080/products'
```
그러니 이런 결과가 돌아갑니다. 이 결과를 더 예쁘게 가공해 쓸 수도 있겠죠.
![image](https://user-images.githubusercontent.com/1855714/108852525-11d76e80-7629-11eb-8840-1147c73d142f.png)

작업하신 상태에서는 `handleProductBadRequestException`가 이걸 가로채서 다음 결과만 나옵니다.

```
HTTP/1.1 400 
Content-Type: application/json
Transfer-Encoding: chunked
Date: Tue, 23 Feb 2021 13:48:14 GMT
Connection: close

{"message":"Product bad request"}
```


### Service 에서의 예외

다른 평범한 Java 애플리케이션을 작성하는 것과 크게 다를 바 없습니다. 예외를 던져야 한다면 예외를 던지고, 예외를 받는(`catch`) 녀석은 예외를 스스로 해결하던가, 또는 자신의 호출자에게 한 번 더 던질 것인가를 결정해야 합니다. `Service 내에서는 예외를 잡는 부분이 없는 것 같`다면 Service 내 코드에서 예외를 던지고 잡는 코드를 추가하면 됩니다. 예외가 던져지는 곳이 서비스인지 컨트롤러인지는 별로 중요하지 않습니다.

### `@DisplayName`

이건 정답이 없어서 상황에 맞는 더 좋은 메시지를 항상 고민하는 수 밖에 없습니다.
`만약 저징되어 있는 고양이 장난감의 아이디와 업데이트 될 이름, 메이커, 가격, 이미지가 주어진다면` 같은 메시지는 저도 참 고민스럽네요. 저라면 테스트 메시지를 조금 더 추상화할 것 같아요. 테스트 설명은 너무 추상적이어도 안 되지만 너무 구체적이어도 좋지 않습니다.

### `@NotNull`, `@NotBlank`

이 테스트는 제 경우엔 Http Rest 테스트로 처리하곤 합니다. 또는 예외가 던져지는지 정도까지만 확인해요.
그리고 `blank일 때`, `null일 때`를 나누는 것은 테스트 코드 양이 상당히 늘어나고, 작성하는 사람도 굉장히 피곤하죠. 이런 경우엔 배열이나 `List`를 사용해서 테스트 코드를 회전시키는 방법을 권합니다. (이번주에 리뷰하면서 도와드릴게요).

### `jacoco` `toString()`
어느 클래스에 있는 `toString` 메소드인가요? 모든 클래스의 `toString` 메소드가 다 그런가요?

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -23,21 +24,33 @@ public ProductService(ProductRepository productRepository) {
     public Product getProduct(Long id) {
-        return findProduct(id);
+        return productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
```
여기에서 던진 예외를 컨트롤러가 `catch`로 받아주거나 컨트롤러에 예외 핸들러가 있어야겠군요.

sunggyupaik replied:
```java
@GetMapping("/{id}")
public Product detail(@PathVariable Long id) {
  try {
      return productService.getProduct(id);
  } catch(ProductNotFoundException e) {
      throw new ProductNotFoundException(id);
  }
}
```
Service에서 `ProductNotFoundException`을 발생했을 때, 마찬가지로 `Controller`에서 예외를 발생시키도록 했습니다. 
그런데, 이정도로만 나타내기에는 조금 허전한 느낌이네요, 예외를 던질 때 어떤 내용 더 추가하면 좋을까요?

johngrib replied:
상황에 적합한 에러 메시지를 작성해보시면 좋을 것 같아요. 에러 메시지는 유지 보수를 하는 성규님 자신을 위해 작성한다고 생각하시면 됩니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -17,122 +24,264 @@
+    @DisplayName("getProducts 메서드는")
+    class Describe_getProducts {
+        @Nested
+        @DisplayName("만약 고양이 장난감 목록이 존재한다면")
```
`고양이 장난감`이라는 명칭과 `Product`라는 명칭 사이의 갭을 생각해 봅시다. `Product`쪽이 더 추상화가 되어 있는 이름이기 때문에 두 단어의 추상화 레벨에 갭이 생깁니다. 이 갭을 줄일 필요가 있습니다. 지금은 고양이 장난감만 팔고 있는 작은 인터넷 쇼핑몰이지만, 장사가 아주 잘 된다면... 그래서 꽤 고객이 더 늘어난다면... 사장님이 어느날 이런 요청을 해오실 수도 있습니다.

"내일부터 고양이 간식도 판매합니다."

그렇다면 소스코드는 전혀 고치지 않아도 돌아갈 수도 있지만, 주석은 고쳐야 할 수도 있습니다.

sunggyupaik replied:
### 고양이 장난감
1. `catToy`
2. `catProduct`

### 고양이 간식
1. `catFeed`
2. `catFood`

이렇다면 좀 분리가 가능할 것 같습니다

johngrib replied:
저는 주석을 `상품`으로 고치는 방향을 생각했는데요, 성규님의 선택도 하나의 방법이라고 생각합시다. 그렇게 가봅시다!

sunggyupaik replied:
네, 그런데 아마 테스트때문에 `Product`로 유지하고 `상품`으로 이어가는게 낫겠습니다

sunggyupaik committed:
Add PATCH /user/{id} test in controller, service

Add update Test. Revise Create test

sunggyupaik committed:
Add PATCH /user/{id} test in controller

sunggyupaik committed:
Add delete method in Service, Controller

Add delete method. Add verify in update controller test

sunggyupaik committed:
Add POST /user service test, repository

Add createUser test in service.
Add UserRepository, JpaUserRepository

sunggyupaik committed:
Add given, verify test in create service test

sunggyupaik committed:
Add PATCH /user/{id} service test, service

Add updateUser() in service test. Revise
updateUser() in service

sunggyupaik committed:
Add assertThat at create, update in service test

Make assertThat check in cerate, update method in service test

sunggyupaik committed:
Add DELETE /user/{id} test in service

sunggyupaik committed:
Add try-catch in controller, Change valid Message

Add try-catch at detail(), delete() in controller
Change MethodArgumentNotValidException

sunggyupaik committed:
Add ProudctBadReuqest at create() in service

sunggyupaik commented:
답변 너무 감사드립니다~!!! 짧은 시간이지만 해당 피드백에 대해서 나름 생각해보았습니다.

## Service에서의 예외 , `@NotNull`


```java
public Product createProduct(ProductData productData) {
        if(productData.getName().isBlank())
            throw new ProductBadRequestException("name");

        if(productData.getMaker().isBlank())
            throw new ProductBadRequestException("maker");

        if(productData.getPrice() == null) {
            throw new ProductBadRequestException("price");

        ...
        }
```
`Service`에서 위와 같은 형태의 예외 처리는 어떨까요? `Service`에서 직접 만든 `ProductBadRequestException.class`를 이용해서 예외를 발생시켜보았습니다.   `Controller`단에서는 `@Valid`로 예외처리가 되기 때문에 여기서 발생한 예외를 `Controller`에서는 따로 반응하도록 하지 않았습니다. 그런데 과연 `Controller`에서 `MethodArgumentNotValidException`로 예외를 처리하고 `Service`에서 `ProductBadRequestException`로 예외를 처리하는 것이 크게 차이가 없는지 궁금합니다.

`ProductBadRequestException.class` 는 다음과 같이 만들었습니다.
```java
public ProductBadRequestException(String variable) {
    super(String.format("%s 값은 값은 필수입니다", variable));
}
```
이제 테스트를 작성합니다
```java
@Nested
@DisplayName("만약 이름이 비어있는 고양이 장난감 객체가 주어진다면")
class Content_WithCatToyWithOutName {
private ProductData createSource;

@BeforeEach
void setUp() {
    createSource = ProductData.builder()
            .name("")
            .maker(CREATED_PRODUCT_MAKER)
            .price(CREATED_PRODUCT_PRICE)
            .imageUrl(CREATED_PRODUCT_IMAGEURL)
            .build();
}

@Test
@DisplayName("이름값이 필수라는 메세지를 응답한다")
void itReturnsBadRequestMessage() {
   assertThat(createSource.getName()).isBlank();
    assertThatThrownBy(() -> productService.createProduct(createSource))
            .isInstanceOf(ProductBadRequestException.class)
            .hasMessageContaining("name 값은 필수입니다");
}
}
```
보통 입력값을 넣지 않고 생성을 눌렀을 때, `""` 와 `null`중에 어떤 값이 입력값으로 들어오는지 모르겠지만 이걸 알면 테스트할 때 한층 편할 것 같습니다.

`assertThat(createSource.getName()).isBlank()` 이런 테스트가 맞는지 고민이 됩니다.

## @Valid 예외 메세지
```java
   public ErrorResponse handleProductBadRequestException(MethodArgumentNotValidException e) {
        String message = e.getBindingResult()
                .getAllErrors()
                .get(0)
                .getDefaultMessage();

        return new ErrorResponse(message);
    }
```
위와 같이 처리하면 아래와 같은 에러 메세지를 얻을 수 있습니다. 저번에 `@NotBlank(message = "이름 값은 필수입니다")` 라고 처리한 것을 `getBindingResult()`로 꺼내서 박았습니다.
![image](https://user-images.githubusercontent.com/45138206/108998086-bc13cc80-76e3-11eb-9993-f8ab4f785071.png)

그런데 계속 에러 메시지를 보면서 좀 더 구체적 내용이 있으면 좋겠다는 생각을 했습니다.
지금 `ErrorResponse`클래스에서는 단순히 변수로 `String message`정도 밖에 가지고 있지 못합니다.
이 클래스에 `상태코드`같은 내용도 더 추가하나요?
혹시 예외 상황에서 어떤 값들을 보여주는지 어느정도의 가이드라인이 있을까요?

## `@DisplayName`
`만약 저징되어 있는 고양이 장난감의 아이디와 업데이트 될 이름, 메이커, 가격, 이미지가 주어진다면`

- 만약 저장되어 있는 고양이 장난감의 아이디와 업데이트 될 객체가 주어진다면
- 만약 저장되어 있는 아이디와 객체가 주어진다면
- 만약 업데이트 될 고양이 장난감 객체가 주어진다면
- 만약 고양이 장난감 객체가 주어진다면
- 만약 유효한 객체가 주어진다면

`적절하게`라는 단어가 제일 어렵네요 ㅋㅋ `적절한` 테스트 설명문과 변수명, 메서드명....

sunggyupaik committed:
Add test without name in service

sunggyupaik committed:
Add ProductBadRequestExcetion in Advice

johngrib reviewed (commented):
## Service에서의 예외, `@NotNull`

테스트를 돌려 보시니 어떤 결과가 나온던가요?

한편 예외 처리를 할 때 주의해야 할 점은 예외 처리엔 오답은 있어도 정답은 없다는 것(...)인데요, 다양한 방법을 실험해보시고 장단점을 직접 체득하는 것이 중요합니다. 회사마다/팀마다 예외 처리 방식이 조금씩 다르기 때문입니다. 보통은 Spring을 최대한 활용해서 예외를 필요한 한도 내에서 최소한으로 작성하는 걸 선호하는 경향이 강하긴 해요. 하지만 `필요한 한도 내에서 최소한으로` 같은 것은 주관적인 개념이기 때문에 이것 또한 답이 없기는 마찬가지입니다.

`assertThat(createSource.getName()).isBlank()` 전 이 부분이 불필요한 것 같아요.

## `@Valid` 예외 메시지

상태 코드는 제일 윗줄에 있습니다. `HTTP/1.1 400`. 그러니 예외 클래스에 굳이 더 추가할 필요는 없겠죠. 여기에 뭘 더 추가하는지는 디버깅하는 사람과 모니터링하는 사람을 위한 서비스 정보라 생각하시면 됩니다. 무엇을 더 추가하면 좋을까요? 예외 메시지를 어떻게 작성하고 어떤 내용을 쓰는지도 회사와 팀에 따라 천차만별이긴 합니다. 왜냐하면 로그를 무엇을 남길 것인가? 하는 문제가 걸려 있기 때문입니다. 이것도 답이 없지요. 지금 생각나는 문서가 하나 있는데요, [RFC 5424]( https://tools.ietf.org/html/rfc5424 ) 입니다. 저도 예전에 읽다가 끝까지 읽지는 못했어요. 이 문서에 바로 Syslog Message Severities가 정의되어 있습니다.

![image](https://user-images.githubusercontent.com/1855714/109008114-48c48780-76f0-11eb-889f-2faed8a4d8b6.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserService {
```
서비스에 JavaDoc 주석을 작성해 주세요.

sunggyupaik replied:
네, 오늘은 JavaDoc 주석을 달아보겠습니다

sunggyupaik replied:
주석의 예시를 살펴보았는대요, 예를 들어 아래의 `valueOf`와 `copyValueOf`는 
`@param` 에서 주석다는 방법,
콤마를 찍는 것, 
`{@code String}`  등으로 예외를 정하는 법 등 다양한 방법이 있습니다.
이는 작성자의 취향차이로 봐도 될까요?

![image](https://user-images.githubusercontent.com/45138206/109151892-19745000-77ae-11eb-8c2a-b47064fa9884.png)

![image](https://user-images.githubusercontent.com/45138206/109151912-1f6a3100-77ae-11eb-9f27-a8b70a1de197.png)

johngrib replied:
네 취향 차이입니다. 보통은 팀마다, 회사마다 주석 스타일도 달라서 그때그때 자신의 스타일을 포기하고 맞춰야 하는 경우도 많아요.

sunggyupaik replied:
네, 그렇다면 제 과제에서는 자유롭게 열린마음으로 작성하겠습니다 ㅎㅎ

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -23,9 +25,13 @@ public ProductController(ProductService productService) {
-        return productService.getProduct(id);
+        try {
+            return productService.getProduct(id);
+        } catch(ProductNotFoundException e) {
```
`catch(ProductNotFoundException e)` 구문이 두 번 이상 사용되고 있군요. 이걸 묶어서 처리하는 것도 고려해볼만 하겠어요.

sunggyupaik replied:
https://github.com/CodeSoom/spring-week5-assignment-1/pull/10#discussion_r581033472
`ProductNotFoundException`에 대해 저는 아래와 같이 `@ControllerAdvice`로 핸들러를 만들어서 대응하고 있습니다.
```
@ResponseBody
@ResponseStatus(HttpStatus.NOT_FOUND)
@ExceptionHandler(ProductNotFoundException.class)
public ErrorResponse handleProductNotFound(ProductNotFoundException e) {
    return new ErrorResponse(e.getMessage());
}
```
그렇다면 결국 `Service`에서 예외가 처리되니까  `Controller`에서 `try-catch`를 사용하거나 따로 예외처리를 하지 않아도 되지 않나요?

sunggyupaik replied:
그리고 중복되는 부분을 제거해 보려고 했습니다.
```java
try {
       return productService.getProduct(id);
       } catch(ProductNotFoundException e) {
            throw new ProductNotFoundException(id);
}
```
이 형식에서 `throw new ProductNotFoundException(id);` 여기를 수정하고 싶은데, 이 클래스의 반환형이 `Product`이다보니, 어떤 방식으로 다시 catch에서 던저줘야 하는지 모르겠습니다.

johngrib replied:
> 그렇다면 결국 `Service`에서 예외가 처리되니까 `Controller`에서 `try-catch`를 사용하거나 따로 예외처리를 하지 않아도 되지 않나요?

그렇습니다! 제가 기다리던 질문입니다. 예외를 반드시 컨트롤러에서만 처리해야 한다고 믿는 사람들도 있는데요, 사실도 아니고 바람직하지도 않은 믿음입니다. 우리는 여기에서 컨트롤러 레이어, 서비스 레이어로 이루어진 여러 계층의 소프트웨어를 작성하고 있는 셈인데요. 어떤 예외는 레이어를 넘나들고, 어떤 예외는 레이어 내에서 다른 레이어에 공개되지 않도록 조용히 처리해야 합니다. 두 예외를 잘 구분해서 처리해야겠죠. 핵심은 이것입니다. "서비스/컨트롤러 입장에서 외부에 공개해야 하는 예외 상황이란 무엇이며 외부에 알리지 않고 조용히 처리할 예외는 무엇인가?"

여기에서 또다시 취향이 갈리고, 팀 컨벤션이 생기고 하는 영역이 됩니다. 제 취향은 가급적이면 예외를 던지지 않고 예외가 발생한 곳에서 잘 처리하는 것이긴 한데요, 어떤 사람은 제 의견에 크게 반대하기도 할 것입니다.

다음에 대해서도 잘 생각해 봅시다.
- 동영상 파일에 0.1초 정도 손상된 구간이 있다면 동영상 재생 프로그램은 사용자에게 예외를 알리지 않고 0.1초 정도 까만 화면을 보여주고(알아서 예외를 처리하고) 계속 다음 장면을 재생한다.
- 프린터기로 뭔가를 출력할 때 종이가 부족하면 프린터기는 기계에도 빨간 불을 켜고, 프린터 출력 프로그램에도 종이가 없다는 것을 알리고, 모니터에도 종이가 없다는 경고 메시지를 띄운다.

johngrib replied:
> 그리고 중복되는 부분을 제거해 보려고 했습니다.
> 
> ```java
> try {
>        return productService.getProduct(id);
>        } catch(ProductNotFoundException e) {
>             throw new ProductNotFoundException(id);
> }
> ```
> 
> 이 형식에서 `throw new ProductNotFoundException(id);` 여기를 수정하고 싶은데, 이 클래스의 반환형이 `Product`이다보니, 어떤 방식으로 다시 catch에서 던저줘야 하는지 모르겠습니다.

이건 한 걸음 물러나서 api 관점에서 보시면 좋겠습니다. 이 메소드는 다음 http 요청을 처리하고 있죠?
```
DELETE /products/{id}
```

여기에서 상품이 없는 경우를 예외로 표현하시고 싶은 거죠. 그런데 URI는 자원 식별자입니다. 우리 컴퓨터 하드디스크의 파일 경로랑 같은 개념입니다. 그리고 여기에서는 `/products/{id}`를 삭제하려고 했는데 이 자원은 없었던 거죠. 이럴 땐 `404 NOT FOUND`를 응답하는 걸 생각할 수 있습니다.

sunggyupaik replied:
이번 프로젝트의 CRUD 예외처리에 대해서 간략히 정리해봤습니다.

`Controller`에서
`조회`, `삭제` : `id`는 `Controller`에서 존재여부 확인이 불가능해서 `Service`에서 예외처리
`생성` , `수정` : `@Valid` 덕분에 유효성 검사가 가능해서 Controller에서 예외처리
혹시 `Controller`에서 `@Valid`를 삭제하면 `Service`에서 모든 예외처리가 가능합니다.

하지만 그렇게 하지 않은 2가지 이유는 다음과 같습니다.
1.굳이 단순한 `생성` `수정` 예외처리가 `Controller` 대신에 `Service`에서 처리 되는 것이 큰 장점이 있는지 모르겠다.
2. `@Valid`는 아래와 같은 이유로 `Service`에서는 사용할 수 없기 때문에 예외처리를 `Service`에서 하는 것은 더 코드가 길다. ( 물론 설계나 효율이 좋다면 그렇게 한다. )

*참고
`@Valid`는 `@Controller`에서만 정상적으로 작동하기 떄문에 `@Service`에서는 사용 할 수 없다. 따라서 `@Service`에서 처리하는 것은 아래와 같이 한땀 한땀 해주어야 한다.

```java
if(productData.getName().isBlank())
    throw new ProductBadRequestException("name");

if(productData.getMaker().isBlank())
    throw new ProductBadRequestException("maker");

if(productData.getPrice() == null) {
    throw new ProductBadRequestException("price");
}
```

그런데, `@Valid` 유효성 검사를 하여 `Controller`에서 예외를 튕겨내면, `Service`에서 테스트할 때, **유효하지 못한 `ProductData`에 대한 예외를 작성할 방법이 없습니다.** `Controller`에서 예외를 잡아내버리기 때문에 `Service`에서는 테스트할 수 있는 장치가 없네요. 한번 더 생각하면 어차피 `Controller`에서 예외 잡는데, `Service`에서 예외를 대해서 테스트 검증을 해야할 필요가 있나 생각이 듭니다.

이 프로젝트에 대한 예외처리 이야기는 여기까지 하고 중요한 것은 아래의 경우 일 것입니다.

> "서비스/컨트롤러 입장에서 외부에 공개해야 하는 예외 상황이란 무엇이며 외부에 알리지 않고 조용히 처리할 예외는 무엇인가?"'

 동영상의 예시(`NullPointerException` ?)는 조용히 처리하는 것일테고, 복사의 예시(`IOException` ?)는 외부에 알리는 것이겠네요. 단순하게 정리해서 조용히 처리하는 것은 로직 상에서 이 예외가 크게 방해가 되지 않아 무시하고 진행할 수 있는 경우, 외부에 알리는 것은 정말 해결하지 않으면 더이상 진행이 안되는 경우로 보입니다.

 또 이것이 내가 예외를 만들 때 `checked`를 사용할 것인가, `unchecked`를 할 것인가에 대한 것과 연관이 있을까요?
( _`checked`는 사용자에게 문제를 해결할 기회를 주는 것이면서, 예외처리를 강행합니다. 하지만 `unchecked`는 사용자가 API의 규칙을 무시하고 사용하거나 예외 상황에서 프로그램을 종료하도록 합니다._  라고 정리해보았습니다.) 이번 프로젝트도 그렇고, 우리가 흔히 작성하는 이 `RuntimeException` 즉, `unchecked`가 조용하게 처리할 수 있는 예외이고 `checked`는 외부에 알리는 예외로 비유할 수 있는지 생각해보았습니다.

 `Delete /products/{id}` 경우 `Service`에서 `ProductNotFoundException`예외를 던지고 `ControllerAdvice`에서 이 예외를 잡아서 처리하기 때문에, 굳이 `Controller`에서 처리하지 않아도 될 것 같습니다.

johngrib replied:
`@Valid는 @Controller에서만 정상적으로 작동하기 떄문에 @Service에서는 사용 할 수 없다.`는 사실이 아닙니다. 서비스에서도 사용할 수 있는데 좀 더 귀찮죠.

`id`를 컨트롤러에서 존재여부 확인이 불가능...하지는 않습니다. 리턴 타입에 주목하면 해결할 수 있는 문제입니다. 이 문제의 가장 널리 알려진 해결책은 `Optional`을 쓰는 거죠.

`checked exception`과 `unchecked exception`에 대해 공식 문서를 읽어 보시기를 권해드립니다.
https://docs.oracle.com/javase/specs/jls/se15/html/jls-11.html#jls-11.1.1

sunggyupaik replied:
잘못된 부분에 대해서 시간을 투자하고 바로 잡도록 하겠습니다. 
그리고 전체적으로 시간에 쫓기어 산만하게 글을 썼었네요

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -1,19 +1,15 @@
-import lombok.Builder;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
`import *`을 사용하지 않는 습관을 들입시다.

sunggyupaik replied:
`lombok`은 자동으로 변하나봅니다. 하나씩 풀었습니다

johngrib replied:
이거 인텔리제이 설정으로 `*` 안 쓰게 할 수 있어요.

yangseungin replied:
안녕하세요. 동기분들 리뷰를 한번씩 보다가 저도 동일한 경험을 한적이 있어서 설정법을 기록해놨었습니다.
기본값으로 5개가 넘어가는경우 *로 묶어버리더라구요
https://giantdwarf.tistory.com/58

sunggyupaik replied:
넵, 도움주신 양승인님도 감사드립니다 ㅎㅎ

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,38 @@
+@Entity
+@NoArgsConstructor
+@Getter
+@Setter
```
이렇게 하면 `id`에도 setter가 생기는 문제가 발생합니다.

sunggyupaik replied:
리팩토링 하겠습니다~!

johngrib commented on app/src/test/java/com/codesoom/assignment/dto/ErrorResponseTest.java:
```diff
@@ -0,0 +1,21 @@
+        assertThat(errorResponse.getMessage()).isEqualTo("Not Found");
+    }
+
+}
```
파일 마지막에 개행 문자를 추가해 주세요.

sunggyupaik replied:
넵 수정했습니다!

yangseungin replied:
Preferences - Editor - General - Save Files에서 Ensure an empty line at the end of a file on save를 체크하면 저장시 파일의 마지막줄에 개행문자가 자동으로 추가됩니다.

sunggyupaik replied:
넵 ㅎㅎ 양승인님 다시한번 감사드려요~!

sunggyupaik committed:
Change with FeedBack

Remove import lombok.*;
Remove @Setter in User.class
Add empty line at last in ErrorResponseTest
Remove create-isBlank() test in ProductServiceTest

sunggyupaik committed:
Make UserData, Revise Service, Controller Test

Make update(), toEntity() in UserData,
Revise create, update test from User to UserData

sunggyupaik committed:
Revise test setUp about UserData

sunggyupaik committed:
Add JavaDoc at ProductService, UserService

sunggyupaik committed:
Revise Java Docs

Add {@code id}, change @throws content

sunggyupaik committed:
Add javaDocs to all classes

sunggyupaik committed:
Add Dozer library

Add Dozer, change toEntity() -> mapper at create in user service

sunggyupaik committed:
Add Dozer at create in Product service

sunggyupaik committed:
Add Dozer mapper at update about Product, User

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -211,6 +211,54 @@ void itReturnsBadRequestMessage() {
             }
         }
+
+        @Nested
+        @DisplayName("만약 메이커값이 비어있는 고양이 장난감 객체가 주어진다면")
+        class Content_WithCatToyWithOutMaker {
+            private ProductData createSource;
+
+            @BeforeEach
+            void setUp() {
+                createSource = ProductData.builder()
+                        .name(CREATED_PRODUCT_NAME)
+                        .maker("")
+                        .price(CREATED_PRODUCT_PRICE)
+                        .imageUrl(CREATED_PRODUCT_IMAGEURL)
+                        .build();
+            }
```
훌륭한 컨텍스트 명세 구현입니다.

sunggyupaik replied:
계속 사용하다보니 좀 더 눈과 손에 익네요 ㅎㅎ

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -211,6 +211,54 @@ void itReturnsBadRequestMessage() {
+            void itReturnsBadRequestMessage() {
+                assertThatThrownBy(() -> productService.createProduct(createSource))
+                        .isInstanceOf(ProductBadRequestException.class)
+                        .hasMessageContaining("maker 값은 값은 필수입니다");
```
저는 메시지까지 테스트할 필요는 없다고 생각하지만, 메시지가 중요하다면 이렇게 하는 것도 좋겠죠.

sunggyupaik replied:
제가 계속 예외처리 메세지(?)에 좀 집착하는 느낌이 없지않아 있네요.. ㅎㅎ

johngrib replied:
좋은 습관이에요. 다만 항상 트레이드 오프를 고려할 수 있으면 더 좋습니다.

sunggyupaik replied:
네, 변수,메세지, 처리방식의 변화에 좀 더 민감해지겠군요

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/UserRepository.java:
```diff
@@ -4,10 +4,16 @@
 import java.util.Optional;
 
+/**
+ * 유저를 조회, 저장, 삭제하는 저장소이다.
```
이렇게 주석을 작성하면 새로운 기능을 추가하거나 기능 하나가 빠지게 되면 주석도 바꿔야 하겠죠?

sunggyupaik replied:
네, 좀 더 추상화 시키기 위해 메서드에 대한 정보는 삭제하겠습니다

johngrib reviewed (changes requested):
PR에 `.class` 바이너리 파일이 잔뜩 들어있습니다. `git rm`으로 제거하고 PR을 갱신해 주세요.

sunggyupaik commented:
> PR에 `.class` 바이너리 파일이 잔뜩 들어있습니다. `git rm`으로 제거하고 PR을 갱신해 주세요.

어허... 이번에도 또 섞여들어갔군요... 저번에는 그냥 지나쳤지만 이번에는 해결하였습니다

sunggyupaik committed:
Revise feedback

sunggyupaik commented:
오늘은 코드에 대한 수정이 거의 없습니다. ㅎㅎ 답글만 확인해주시면 감사하겠습니다~

sunggyupaik committed:
Change Test @Display, Add Exception Message

Change ProductServiceTest,ProductControllerTest, UserServiceTest

sunggyupaik committed:
Delete domain logic, Change @DisplayName about product

johngrib commented on .gitignore:
```diff
@@ -24,3 +24,11 @@ Thumbs.db
 ehthumbs.db
 ehthumbs_vista.db
+
+#Binary files
+.class
+<<<<<<< HEAD
+
+app/out
+=======
+>>>>>>> 3c21da3c3d83cba6fe80a76ed4624de6f73a313b
```
여기에서 git 머지나 리베이스 중 conflict가 발생했던 모양이네요. 불필요한 내용을 삭제해주시면 될 것 같습니다.

sunggyupaik replied:
네,, 확인했습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    /**
+     * 주어진 식별자 해당하는 사용자를 리턴한다.
```
조사가 빠졌습니다. JavaDoc의 summary description은 API 문서 웹 페이지로 빌드될 것을 생각하고 가능한 한 잘 작성해야 합니다.
```suggestion
     * 주어진 식별자에 해당하는 사용자를 리턴한다.
```

sunggyupaik replied:
네 명심하겠습니다
https://github.com/CodeSoom/spring-week5-assignment-1/pull/10#discussion_r584255954

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,91 @@
+     *
+     * @param id - 조회하려는 사용자의 식별자
+     * @return 주어진 {@code id}에 해당하는 식별자
+     * @throws UserNotFoundException 만약
+     *         {@code id}가 저장되어 있지 않은 경우
```
다음과 같이 작성하면 어떨까요?
```suggestion
     * @throws UserNotFoundException 만약
     *         {@code id}에 해당되는 사용자가 저장되어 있지 않은 경우
```

sunggyupaik replied:
설명문 작성은 정말 여러번 가다듬었다고 확인했는데도 아쉬운 부분이 많네요 ㅎㅎ
꼼꼼하게 봐주신것처럼 저도 지속적으로 저의 코드를 꾸준하게 확인하도록 습관을 가지겠습니다.

johngrib replied:
성규님은 노력하시는 것이 늘 눈에 보이고, 성장폭도 2주마다 훅훅 뛰는 것 같아 기쁩니다. 화이팅!

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,49 @@
+    @NotBlank(message = "name 값은 필수입니다")
+    private String name;
+
+    @Mapping("email")
+<<<<<<< HEAD
+=======
+    @NotBlank(message = "email 값은 필수입니다")
+>>>>>>> 3c21da3c3d83cba6fe80a76ed4624de6f73a313b
+    private String email;
```
여기에서 conflict를 해결해 주세요. 이 부분 때문에 프로젝트 전체가 컴파일되지 않습니다.

sunggyupaik replied:
흐,, 다음날 알았습니다 해결했습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @param userData - 새로 저장하고자 하는 사용자
+     * @return 저장 된 사용자
+     * @throws MethodArgumentNotValidException 만약 주어진 사용자의
+     *         이름이 비어있거나, 이메일이 비어있거나, 비밀번호가 비어있는 경우
```
이 `@throws`는 삭제해도 괜찮을 것 같습니다. 컨트롤러의 책임이 아닙니다.

sunggyupaik replied:
네, `Controller`의 책임이 아니네요~!

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -20,176 +27,384 @@
-        mockMvc.perform(
-                post("/products")
-                        .accept(MediaType.APPLICATION_JSON_UTF8)
+                mockMvc.perform(post("/products").accept(MediaType.APPLICATION_JSON_UTF8)
```
`MediaType.APPLICATION_JSON_UTF8`은 deprecated 입니다. 다른 것으로 바꿔 주세요. JavaDoc을 잘 읽으면 무엇으로 바꿔야 하는지도 알 수 있습니다.

![image](https://user-images.githubusercontent.com/1855714/109299127-f4e4aa80-7877-11eb-9f05-23a6d1d0b033.png)

sunggyupaik replied:
네, deprecated되었다고 글씨가 그어져 있음에도 간과했었습니다
![image](https://user-images.githubusercontent.com/45138206/109411420-4b232c00-79e5-11eb-9acc-7a179004d403.png)

`mockMvc` 초기에 필터를 통해 `"UTF-8`"을 응답으로 설정하였습다

sunggyupaik committed:
Change Service

sunggyupaik committed:
Change conflict in UserData, .gitignore

sunggyupaik commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,91 @@
+     * 주어진 식별자 해당하는 사용자를 리턴한다.
+     *
+     * @param id - 조회하려는 사용자의 식별자
+     * @return 주어진 {@code id}에 해당하는 식별자
```
고치다보니 바로 위에 잘못 작성한 부분을 발견했네요!

sunggyupaik commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * 주어진 식별자 해당하는 사용자를 리턴한다.
+     *
+     * @param id - 조회하려는 사용자의 식별자
+     * @return 주어진 {@code id}에 해당하는 식별자
```
수정하던 중 바로 위에 잘못된 주석을 하나 더 발견했네요~!
`@return`은 식별자가 아니라 사용자일텐대요

johngrib replied:
이렇게 자신의 PR에 스스로 리뷰를 남기는 것도 좋은 습관입니다 :-)

sunggyupaik commented:
제가 어제 시간을 계속 투자했던 부분이 `createUser`와 `updateUser`에 매개변수에 대한 유효성이었습니다.
`npm test` 테스트 중에 `update`에서는 `name`, `password`만 가지고 수정하는 부분이 있었습니다.
![image](https://user-images.githubusercontent.com/45138206/109411936-a1459e80-79e8-11eb-945b-20d43262ad66.png)



그런데 저는 'create`에서와 같은 유효성 검사를 사용했기 때문에, `email`이 없다는 이유로 계속 `400 Bad Requset` 에러를 냈었습니다. 그래서 각 `Controller`요청에 따라 각각의 Dto를 만들어야겠다는 것을 뼈저리게 느꼈습니다.
(해당 테스트케이스는 과제 해설 영상을 보지 않고서는 예측하기 어려웠던 부분이었습니다 ㅜㅜ)

그리고 하나 나아가서, 지금 저는 반환형을 `User`로 잡고 있는데 종립님은 이 부분도 `Dto`로 처리하는 것이 안전하다고 생각하시나요?

sunggyupaik commented:
종립님은 시간관리와 체력관리 어떻게 하시는지 여기에 물어봐도 될까요? ㅋㅋ

 저는 금요일밤부터 긴장이 너무 풀려서 토요일, 일요일에 더욱 알찬 공부에 잘 집중이 되지 않습니다... 주말에 해야할 것에 대해 명확히 목표를 잡아두고 강하게 동기부여가 필요할까요? 개인적 일정으로 주말에 많은 시간을 공부에 투자할 수 없기도 하지만, 자투리 시간동안 더 효율적으로 이용하고 싶은데 저항성이 평일보다는 몇배나 더 크네요... 

 현재는 가정도 있으시고 코드숨 멘토님까지 같이 하실텐대 일정관리를 어떻게 하시는지 궁금하고 조언 구하고 싶습니다. 종립님은 내일 휴일도 또 새로운 공부를 하고 계시겠죠!? 저도 달립니다!

 또 운동도 하셨다고 하는데, 평소에도 꾸준하게 관리하시나요? 1월부터 일 시작하고 그냥 매일 이렇게 살다가 몸이 못버티겠구나 싶습니다. 헬스라도 다녀야하는지 하다못해 유산소도 해야되는데ㅋㅋ , 주말에 공부에 대한 정신력이 약해지는게 체력 문제도 있는 것 같구요...

 일하구 개인적 모임, 코드숨까지 하면 2월 정말 순식간에 사라졌네요!! 종립님도 2월 마무리 잘하시구 내일도 잘 쉬시길 바랍니다~!

sunggyupaik committed:
Change @DisplayName, add UserUpdate exception

Change @DisplayName in UserService.test, PorudctService.test
Add UserUpdate badrequestException

johngrib commented:
> 그리고 하나 나아가서, 지금 저는 반환형을 `User`로 잡고 있는데 종립님은 이 부분도 `Dto`로 처리하는 것이 안전하다고 생각하시나요?

상황에 따라 다르긴 한데요, 저는 따로 만들어 처리하는 걸 선호합니다. 경험상 그게 유지보수하면서 제일 골치가 덜 아팠어요.

johngrib commented:
> 종립님은 시간관리와 체력관리 어떻게 하시는지 여기에 물어봐도 될까요? ㅋㅋ
> 
> 저는 금요일밤부터 긴장이 너무 풀려서 토요일, 일요일에 더욱 알찬 공부에 잘 집중이 되지 않습니다... 주말에 해야할 것에 대해 명확히 목표를 잡아두고 강하게 동기부여가 필요할까요? 개인적 일정으로 주말에 많은 시간을 공부에 투자할 수 없기도 하지만, 자투리 시간동안 더 효율적으로 이용하고 싶은데 저항성이 평일보다는 몇배나 더 크네요...
> 
> 현재는 가정도 있으시고 코드숨 멘토님까지 같이 하실텐대 일정관리를 어떻게 하시는지 궁금하고 조언 구하고 싶습니다. 종립님은 내일 휴일도 또 새로운 공부를 하고 계시겠죠!? 저도 달립니다!
> 
> 또 운동도 하셨다고 하는데, 평소에도 꾸준하게 관리하시나요? 1월부터 일 시작하고 그냥 매일 이렇게 살다가 몸이 못버티겠구나 싶습니다. 헬스라도 다녀야하는지 하다못해 유산소도 해야되는데ㅋㅋ , 주말에 공부에 대한 정신력이 약해지는게 체력 문제도 있는 것 같구요...
> 
> 일하구 개인적 모임, 코드숨까지 하면 2월 정말 순식간에 사라졌네요!! 종립님도 2월 마무리 잘하시구 내일도 잘 쉬시길 바랍니다~!

일정관리는 딱히 없고, 저는 매일 일과가 똑같아서 별다른 생각 안 하고 그냥 하던 일을 합니다. 그냥 일과가 단순해요.
제 일과는 밥 먹으면서 TV 보고 -> 다음 밥 먹을때까지 컴퓨터하면서 놀거나 공부하고의 반복인데요... 평일엔 이게 퇴근 후에 한 번 발생하고, 주말엔 3번 발생한다는 차이가 있어요. 동기부여는 잘 모르겠고 그냥 어제 하던대로 해요. 유튜브 많이 본 날은 후회하면서 자고, 공부 많이 한 날은 좀 놀걸 그랬네 하고 후회하면서 잡니다. 동기를 어떻게 갖건 후회하는 건 비슷한 것 같아요.

운동은 시간을 따로 내지 않아요. TV볼 때 운동을 하던가, 점심 시간에 시간이 조금 남을 때 운동을 하던가 합니다.

한 주간 고생 많으셨어요. 성규님도 2월 마무리 잘 하시고 남은 휴일 잘 지내시길 바래요 ㅎㅎ

sunggyupaik commented:
네ㅎㅎ ( 며칠전에 2020년도 마켓컬리 인터뷰 잘봤어용 )

johngrib commented:
제 인터뷰를 봐주셨다니 쑥쓰럽네요. 감사합니다!
